falco.yaml

```diff
@@ -290,6 +291,34 @@ outputs:
   rate: 0
   max_burst: 1000
 
+# [Experimental] `queue_capacity_outputs`
+#
+# Falco utilizes tbb::concurrent_bounded_queue for handling outputs, and this parameter
+# allows you to customize the queue capacity. Please refer to the official documentation:
+# https://oneapi-src.github.io/oneTBB/main/tbb_userguide/Concurrent_Queue_Classes.html.
+# On a healthy system with optimized Falco rules, the queue should not fill up.
+# If it does, it is most likely happening due to the entire event flow being too slow,
+# indicating that the server is under heavy load.
+# 
+# Lowering the number of items can prevent memory from steadily increasing until the OOM
+# killer stops the Falco process. We provide recovery actions to self-limit or self-kill
+# in order to handle this situation earlier, similar to how we expose the kernel buffer size
+# as a parameter.
+# However, it will not address the root cause of the event pipe not keeping up.
+# 
+# `items`: the maximum number of items allowed in the queue, defaulting to 0. This means that
+# the queue is unbounded.
+# You can experiment with values greater or smaller than the anchor value 1000000.
+# 
+# `recovery`: the strategy to follow when the queue becomes filled up. This also applies when
+# the queue is unbounded, and all available memory on the system is consumed.
+# recovery: 0 means continue.
+# recovery: 1 means simply exit (default behavior).
+# recovery: 2 means empty the queue and then continue.
+queue_capacity_outputs:
+  items: 0
```

Comment:
@leogr as per our chat preserving the old default (unbounded and exiting). Obviously we can change the encoding etc.

Reply:
Furthermore, @leogr I would like to better understand the interrelationships with for instance `output_timeout`, `buffered_outputs` and `outputs.rate` and `outputs.max_burst`. Their descriptions are also still more cryptic especially for the outputs one. aka I am looking for a scaling factor recommendation aka if you change let's 

```
outputs:
  rate: 100000
  max_burst: 100000000
``` 

what capacity of the queue should you choose in relation?

Maybe the timeout default is not right? we block for too long? How was the current default derived?

Reply:
Also reading this raises some question marks ... because not having a recovery strategy as well here appears suboptimal:

```
# It's important to note that Falco outputs will not be discarded from the
# output queue. This means that if an output channel becomes blocked
# indefinitely, it indicates a potential issue that needs to be addressed by the
# user.
output_timeout: 2000
```

Reply:
**Why the `output_timeout` is  so important?**

 When output consumer blocks (ie. imagine the `program` output or any I/O related op that happens downstream to Falco) and the alert can be delivered within the given deadline, Falco uses `stderr` to notify the user of a not addressable problem. 
 
 Why `stderr` ? Because output channels are blocked. 
 
Why the problem is not addressable by Falco? Because such a situation indicates an I/O problem downstream to Falco (for example, a network slowdown) or, more rarely, a misconfiguration (for instance, imagine configuring the `program` output with a program that blocks for more than 2 seconds, you can simulate this using `sleep 10`).

So `output_timeout` is just an emergency mechanism to notify the user that something (outside Falco) is blocking Falco and Falco is giving up because it can't handle the problem. It has no direct with Falco's performance or with its behavior during normal operations.

Reply:
Let me read more carefully and check later. Meanwhile would you have scaling / relationship recommendations for the rate and bursts wrt queue capacity? https://github.com/falcosecurity/falco/pull/2711#discussion_r1281348771

Reply:
thank you Leo for the good explanation! 
What about deprecating the `rate_limiter` (just put a deprecation warning in Falco and a `[DEPRACATED]` label in the config) in this release `0.36` and then remove it in `0.37`? WDYT?

Reply:
+1 on this @Andreagit97: 

Don't wanna come across as always being in favor of just getting rid of things, but at the same time, we do need to improve technical clarity and clean things up. Referring to it as "technical debt" without any negative connotation -- it's normal in every project.

@Andreagit97 would you be willing to run point on this? Asking because you know I will be out most of Sep and will limit code contributions now for this release, but happy to help as reviewer!

Reply:
I don't have a strong opinion on the rate limiter. I wonder if it can still be helpful with some use cases or not.
I vaguely recall I had a conversation with @jasondellaluce a while ago.

Anyway, I'm okay with removing it if we get a consensus. In such a case, I recommend putting a deprecation warning for at least 1 release before completely removing it.

Reply:
Hmm I am just confused that we then silently drop events without even counting these drops.  Is there another way to do it? If we keep it can we at least add drop counters in?

I guess it comes back to Falco being a bit of a different type of a service, where the end goal would be to not produce too many alerts and try to never drop an event.

Reply:
yeah I will open an issue to deprecate it and see if we reach consensus